### ISSUE TITLE
Breakage with OpenAL 1.8

### ISSUE THREAD

**Issue by somini** [zevv/bucklespring#50]
First of all, thanks for creating this. I still start buckle on every boot, but muted, so that I can impress people on a touch of a button.

I'm using Arch Linux and a few days ago OpenAL was updated to version 1.18. Suddenly, buckle stops playing sounds and just keeps allocating memory forever and brings the entire computer to a halt.

I used the `-vvv` argument but no logs were printed from the moment it was supposed to play sounds.

**Comment by zevv** [comment]
Hi,

I'll try to reproduce, buy my debian is still at 1.17, so I'll have to see if I can setup an arch VM. Did you try to rebuild buckle after the OpenAL upgrade?

**Comment by somini** [comment]
I did rebuild it but no dice.
I'm available for testing the eventual new version.

**Comment by Libbum** [comment]
Ha! I haven't played around with this for a long while, but wanted to show someone today, only hours after I'd upgraded to OpenAL 1.18 (Arch Linux here too - can confirm issue)!

**Comment by vain** [comment]
Hmm. I updated a few days ago.

```
$ pacman -Qi openal
Name            : openal
Version         : 1.18.0-1
```

Still works fine.

Are you guys *not* using PulseAudio by any chance? (I do.)

**Comment by Libbum** [comment]
I'm using it too, yeah. 

    $ pacman -Qi pulseaudio
    Name            : pulseaudio
    Version         : 10.0-3

**Comment by somini** [comment]
Yes, I'm using pulse audio too.

**Comment by zevv** [comment]
Unfortunately I'm not able to reproduce. Updated to openal1.18, and everything works as before.

We could try to get some insight into what's happening under the hood using strace, but the problem is that OpenAL apps do quite a lot of syscalls all the time, so the output is probably pretty cluttered.

Could you try the following: run

    strace -f ./buckle 2>/tmp/buckle.log

Hit a few keys and exit with ^C right away, and send me the /tmp/buckle.log output?

**Comment by Libbum** [comment]
[buckle.log](https://hastebin.com/isidataton.go)

**Comment by somini** [comment]
My logs are too big for GitHub it seems. `strace -f` produces a ton of logs, without the `-f` option is easier to see what's going on.

I can see that the WAV file is read, closed and then one of the `futex` just hangs for ever.
```
poll([{fd=8, events=POLLIN}], 1, -1)    = 1 ([{fd=8, revents=POLLIN}])
recvmsg(8, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\1\0\v\0\10\0\0\0\0\0\n\0\0\0\0\0\335\16\200\2\0\0\0\0Z\0\3\0d\0\17\0"..., iov_len=4096}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 64
recvmsg(8, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
open("/usr/share/bucklespring/1d-1.wav", O_RDONLY) = 9
fstat(9, {st_mode=S_IFREG|0644, st_size=28158, ...}) = 0
lseek(9, 0, SEEK_SET)                   = 0
read(9, "RIFF\366m\0\0WAVEfmt \20\0\0\0\1\0\1\0D\254\0\0\210X\1\0"..., 4096) = 4096
lseek(9, 4096, SEEK_SET)                = 4096
lseek(9, 4096, SEEK_SET)                = 4096
lseek(9, 24576, SEEK_SET)               = 24576
read(9, "Q\0H\0j\0R\0M\0P\0E\0=\0=\0Q\0>\0006\0M\0002\0F\0;\0"..., 4096) = 3582
lseek(9, 0, SEEK_SET)                   = 0
read(9, "RIFF\366m\0\0WAVEfmt \20\0\0\0\1\0\1\0D\254\0\0\210X\1\0"..., 4096) = 4096
brk(0xde2000)                           = 0xde2000
read(9, "\230\0\37\0)\0\307\0{\0\372\377e\0\310\0\254\0\265\0\27\1\267\0\214\0\0\1\320\0_\0"..., 4096) = 4096
read(9, "y\377i\377b\377H\377[\3774\3775\377q\377\200\377O\377y\377\214\377t\377\236\377\241\377q\377"..., 4096) = 4096
read(9, "$\0\361\377\361\377\370\377\377\377\n\0\364\377\356\377\334\377\332\377\347\377\353\377\353\377\330\377\351\377\312\377"..., 4096) = 4096
read(9, "\351\377\345\377\367\377\n\0\370\377\346\377\314\377\355\377\360\377\351\377\363\377\354\377\336\377\311\377\374\377\10\0"..., 4096) = 4096
read(9, "\2\0\f\0\17\0\t\0\32\0\27\0\364\377\n\0\36\0\2\0\6\0\t\0\n\0\367\377\333\377\377\377"..., 4096) = 4096
read(9, "Q\0H\0j\0R\0M\0P\0E\0=\0=\0Q\0>\0006\0M\0002\0F\0;\0"..., 4096) = 3582
read(9, "", 4096)                       = 0
brk(0xe0d000)                           = 0xe0d000
close(9)                                = 0
futex(0xd54d90, FUTEX_LOCK_PI_PRIVATE, NULL
```

**Comment by somini** [comment]
Here's the same loop with `openal-1.17.2-2`:
```
poll([{fd=8, events=POLLIN}], 1, -1)    = 1 ([{fd=8, revents=POLLIN}])
recvmsg(8, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\1\0\v\0\10\0\0\0\0\0\5\0\0\0\0\0\250}\202\2\0\0\0\0Q\0\2\0Z\0\r\0"..., iov_len=4096}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 64
recvmsg(8, {msg_namelen=0}, 0)          = -1 EAGAIN (Resource temporarily unavailable)
open("/usr/share/bucklespring/1d-0.wav", O_RDONLY) = 9
fstat(9, {st_mode=S_IFREG|0644, st_size=29228, ...}) = 0
lseek(9, 0, SEEK_SET)                   = 0
read(9, "RIFF$r\0\0WAVEfmt \20\0\0\0\1\0\1\0D\254\0\0\210X\1\0"..., 4096) = 4096
lseek(9, 4096, SEEK_SET)                = 4096
lseek(9, 4096, SEEK_SET)                = 4096
lseek(9, 28672, SEEK_SET)               = 28672
read(9, "d\0e\0R\0W\0d\0009\0A\0h\0_\0J\0I\0X\0R\0K\0m\0d\0"..., 4096) = 556
lseek(9, 0, SEEK_SET)                   = 0
read(9, "RIFF$r\0\0WAVEfmt \20\0\0\0\1\0\1\0D\254\0\0\210X\1\0"..., 4096) = 4096
read(9, "\304\3731\374\367\374\275\375w\376E\377\232\377\n\0+\1*\2\246\2\363\2Y\3\304\3\220\3\210\3"..., 4096) = 4096
read(9, "\214\377\334\377\22\0Q\0\221\0\266\0\324\0\t\1\24\1\253\0X\0006\0 \0\t\0\334\377\324\377"..., 4096) = 4096
read(9, "n\0W\0 \0\21\0\330\377\334\377\276\377\256\377\227\377\204\377o\377l\377d\377v\377R\377^\377"..., 4096) = 4096
read(9, "\r\0\31\0\376\377\325\377*\0,\0#\0:\0\35\0008\0?\0R\0X\0J\0D\0,\0"..., 4096) = 4096
read(9, "}\0i\0N\0X\0T\0^\0A\0\17\0/\0O\0O\0V\0?\0005\0.\0\25\0"..., 4096) = 4096
read(9, "\342\377\354\377\361\377\373\377\351\377\327\377\346\377\310\377\365\377\370\377\352\377\334\377\346\377\10\0\374\377\371\377"..., 4096) = 4096
read(9, "d\0e\0R\0W\0d\0009\0A\0h\0_\0J\0I\0X\0R\0K\0m\0d\0"..., 4096) = 556
read(9, "", 4096)                       = 0
brk(0x219f000)                          = 0x219f000
brk(0x2166000)                          = 0x2166000
close(9)                                = 0
futex(0x20ca9b0, FUTEX_LOCK_PI_PRIVATE, NULL) = 0
futex(0x20ca9b0, FUTEX_UNLOCK_PI_PRIVATE) = 0
poll([{fd=8, events=POLLIN}], 1, -1)    = 1 ([{fd=8, revents=POLLIN}])
```

**Comment by vain** [comment]
I see a `/home/tim/.alsoftrc` getting loaded in @Libbum's log. What's the content of that file? Have you tried running without it?

**Comment by Libbum** [comment]
$ cat .alsoftrc 
    hrtf = true
    period_size = 32
    periods = 4

Which are there solely for bucklespring, so the file shouldn't be an issue regardless. But I just checked - no change when the config file is removed.

**Comment by somini** [comment]
Yes, I don't have that file either.

**Comment by blinry** [comment]
I'm on a fully updated Arch Linux, and I'm getting inconsistent behaviour. Most of the time, bucklespring works correctly, but sometimes when I start it, I'm getting no sound. Happy to help to debug this if there's any information which could be useful to you. Bucklespring always appears in the "Playback" tab of pavucontrol.

**Comment by somini** [comment]
I build a newer version of openal with the commit that allegedly fixed this:
https://github.com/kcat/openal-soft/commit/d4f3490a880f779f460332689a396ade69840f50
On Arch, just download the PKGBUILD here:
https://git.archlinux.org/svntogit/packages.git/plain/trunk/PKGBUILD?h=packages/openal
Replace the `_commit=` variable with `d4f3490a880f779f460332689a396ade69840f50` and `makepkg -sri`.

I can confirm this is fixed, so it's only a matter of @kcat releasing a new version.

**Comment by zevv** [comment]
Quoting somini (2017-06-17 00:32:19)

> I build a newer version of openal with the commit that allegedly fixed this:
> https://github.com/kcat/openal-soft/commit/d4f3490a880f779f460332689a396ade69840f50
> On Arch, just download the PKGBUILD here:
> https://git.archlinux.org/svntogit/packages.git/plain/trunk/PKGBUILD?h=packages/openal
> Replace the `_commit=` variable with `d4f3490a880f779f460332689a396ade69840f50` and `makepkg -sri`.
> 
> I can confirm this is fixed, so it's only a matter of @kcat releasing a new version.

Thanks for figuring that out. I'll keep this ticket open for some time
so it is visible to others running into the same problem, we can close
it when the new pkg is available.

--
:wq
^X^Cy^K^X^C^C^C^C

**Comment by compositebowman** [comment]
@somini thanks for specifying the fix. I also ran into this problem in a number of other programs. This issue was referenced and reported to the Arch package maintainer so it is now fixed as of version 1.18.0-2.

**Comment by zevv** [comment]
Thanks for helping out, I'm closing this for now. If anything new pops up, feel free to reopen the issue.
